### PR TITLE
Prevent data race condition  in csi unit tests

### LINKED ISSUE #102116: [Flaky] k8s.io/kubernetes/pkg/volume/csi TestAttacherMountDeviceWithInline
#### Which jobs are flaking:
https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=DATA%20RACE&test=TestAttacherMountDeviceWithInline

- ci-kubernetes-generate-make-test-count10
- pr:pull-kubernetes-unit
- ci-kubernetes-unit
- ci-kubernetes-generate-make-test-count10

#### Which test(s) are flaking:
TestAttacherMountDeviceWithInline

#### Reason for failure:
```
=== RUN   TestAttacherMountDeviceWithInline/stage_unstage_cap_not_set
    csi_attacher_test.go:1393: Running test case: stage_unstage cap not set
W0518 11:07:43.386832  125095 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0518 11:07:43.387265  125095 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
==================
WARNING: DATA RACE
Write at 0x00c002701018 by goroutine 6096:
  k8s.io/kubernetes/vendor/k8s.io/client-go/testing.(*Fake).PrependWatchReactor()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/testing/fake.go:114 +0x435
  k8s.io/kubernetes/pkg/volume/csi.TestAttacherMountDeviceWithInline.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/csi/csi_attacher_test.go:1402 +0x279
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Previous read at 0x00c002701018 by goroutine 5978:
  k8s.io/kubernetes/vendor/k8s.io/client-go/testing.(*Fake).InvokesWatch()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/testing/fake.go:160 +0x207
  k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/storage/v1/fake.(*FakeVolumeAttachments).Watch()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/storage/v1/fake/fake_volumeattachment.go:79 +0x1c4
  k8s.io/kubernetes/vendor/k8s.io/client-go/informers/storage/v1.NewFilteredVolumeAttachmentInformer.func2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/informers/storage/v1/volumeattachment.go:70 +0x23c
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*ListWatch).Watch()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/listwatch.go:111 +0x98
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:414 +0x6dc
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:221 +0x53
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xba
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:220 +0x2b5
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run-fm()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:218 +0x54
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:56 +0x45
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x6d

Goroutine 6096 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1238 +0x5d7
  k8s.io/kubernetes/pkg/volume/csi.TestAttacherMountDeviceWithInline()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/volume/csi/csi_attacher_test.go:1392 +0xa7b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Goroutine 5978 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x70
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:55 +0xc5
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).Run()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:152 +0x4a8
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).Run()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go:410 +0x7c4
==================
    csi_client_test.go:293: calling fake.NodeGetCapabilities for NodeSupportsStageUnstage...
I0518 11:07:43.419875  125095 csi_attacher.go:352] kubernetes.io/csi: attacher.MountDevice STAGE_UNSTAGE_VOLUME capability not set. Skipping MountDevice...
    testing.go:1092: race detected during execution of test
    --- FAIL: TestAttacherMountDeviceWithInline/stage_unstage_cap_not_set (0.04s)
```

#### Anything else we need to know:
- links to go.k8s.io/triage appreciated
- links to specific failures in spyglass appreciated

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind failing-test

#### What this PR does / why we need it:
Prevent data race condition  in csi unit tests

Failing test:
https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=DATA%20RACE&test=TestAttacherMountDeviceWithInline

#### Which issue(s) this PR fixes:

Fixes #102116

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/testing/fake.go [hunk 1]
```diff
@@ -111,6 +111,8 @@ func (c *Fake) AddWatchReactor(resource string, reaction WatchReactionFunc) {
 
 // PrependWatchReactor adds a reactor to the beginning of the chain.
 func (c *Fake) PrependWatchReactor(resource string, reaction WatchReactionFunc) {
+	c.Lock()
+	defer c.Unlock()
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/testing/fake.go:115]: Please run the test TestAttacherMountDeviceWithInline 1000 tests to see if it is still flaky

bazel test --runs_per_test=1000 //pkg/volume/csi/...

Comment (yliaog) [reply staging/src/k8s.io/client-go/testing/fake.go:115]: also what about other Methods, like AddWatchReactor above?

Comment (mengjiao-liu) [reply staging/src/k8s.io/client-go/testing/fake.go:117]: I used the [stress](https://github.com/kubernetes/community/blob/4026287dc3a2d16762353b62ca2fe4b80682960a/contributors/devel/sig-testing/flaky-tests.md#deflaking-unit-tests) command for the unit test:

Here are my test commands:
```
go test ./pkg/volume/csi  -race -c

stress ./csi.test -test.run TestAttacherMountDeviceWithInline 
```

AddWatchReactor also updated.

**Pull request merged**